### ISSUE TITLE
ZFS root pool will not mount if other pools in the zfs cache does not mount

### ISSUE THREAD

**Issue by brydling** [openzfs/zfs#8343]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  Ubuntu
Distribution Version	|  18.04.1
Linux Kernel	|  4.15.0-43-generic
Architecture	|  x86_64
ZFS Version	| 0.7.5-1ubuntu16.4
SPL Version	| 0.7.5-1ubuntu1

### Describe the problem you're observing
I am running Ubuntu 18.04.1 with a luks-encrypted zfs root and have switched from initramfs-tools to dracut to be able to use clevis for automatic decryption using a tang server. This worked fine with only the encrypted root zfs pool in the system.

After adding the encrypted storage pool, which is not decrypted during the initrd stage, the zfs-import-cache.service fails in the initrd stage because the cache now also contains the storage pool (which cannot be mounted since the disks have not been decrypted yet).

My workaround for now is to export the storage pool before generating an initrd image using "dracut -f", so that the storage pool is not included in the zfs cache file that gets included in the initrd image.

### Describe how to reproduce the problem
This is my /etc/crypttab:
```
luks1 UUID=372a6324-d6db-4871-9cc5-3df3afd38935 none luks,discard,initramfs
luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1 /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1 /root/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1.key luks,discard
luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1 /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1 /root/ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1.key luks,discard
```

And these are my pools:

```
# zpool status
  pool: rpool
 state: ONLINE
  scan: none requested
config:

        NAME        STATE     READ WRITE CKSUM
        rpool       ONLINE       0     0     0
          luks1     ONLINE       0     0     0

errors: No known data errors

  pool: tank
 state: ONLINE
  scan: none requested
config:

        NAME                                                     STATE     READ WRITE CKSUM
        tank                                                     ONLINE       0     0     0
          mirror-0                                               ONLINE       0     0     0
            luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1  ONLINE       0     0     0
            luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1  ONLINE       0     0     0

errors: No known data errors
```

### Include any warning/errors/backtraces from the system logs

Boot log:
```
[FAILED] Failed to start Import ZFS pools by cache file.
See 'systemctl status zfs-import-cache.service' for details.
Mounting /sysroot...
[FAILED] Failed to mount /sysroot.
See 'systemctl status sysroot.mount' for details.
[DEPEND] Dependency failed for Initrd Root File System.
[DEPEND] Dependency failed for Reload Configuration from the Real Root.
```

`systemctl status zfs-import-cache.service`:

```
...
Active: failed (Result: exit-code)
...
Process: 773 ExecStart=/sbin/zpool import -c /etc/zfs/zpool.cache -aN (code=exited, status=1/FAILURE)
...
systemd[1]: Starting Import ZFS pools by cache file...
zpool[773]: cannot import 'tank': no such pool or dataset
zpool[773]:         Destroy and re-create the pool from
zpool[773]:         a backup source.
systemd[1]: zfs-import-cache.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: zfs-import-cache.service: Failed with result 'exit-code'.
systemd[1]: Failed to start Import ZFS pools by cache file.
```

`systemctl status sysroot.mount`:

```
...
Active: failed (Result: exit-code)
...
Process: 868 ExecMount=/bin/mount rpool/ROOT/ubuntu /sysroot -t zfs -o zfsutil,rw,relatime,xattr,noacl (code=exited, status=1/FAILURE)
...
systemd[1]: Mounting /sysroot...
mount[868]: filesystem 'rpool/ROOT/ubuntu' cannot be mounted, unable to open the dataset
systemd[1]: sysroot.mount: Mount process exited, code=exited status=1
systemd[1]: sysroot.mount: Failed with result 'exit-code'.
systemd[1]: Failed to mount /sysroot.
```

**Comment by iamjamestl** [comment]
The behavior you describe seems correct to me.  You say the devices for the `tank` zpool are not available when `zfs-import-cache.service` is started because the devices are encrypted, so the `zfs-import-cache.service` rightly fails.  Normally systemd does the right thing because `zfs-import-cache.service` has a dependency on `cryptsetup.target`, but it sounds like you are doing encryption differently.  You should either ensure the devices are unlocked before `zfs-import-cache` is started by expressing that dependency in systemd like:

`/etc/systemd/system/zfs-import-cache.service.d/after-unlock-devices.conf`:
```
[Unit]
After=thing-that-unlocks-devices.service
```

Or if you don't have a systemd service that unlocks the devices, remove `tank` from the default zpool cache so systemd doesn't try to import it at boot: `zpool set cachefile=none tank`.  Then you will need to import it using the `zpool import -d /path/to/devices tank` command.

Or you can set an alternative cache file for `tank`, like: `zpool set cachefile=/etc/zfs/tank.cache tank` and import it with `zpool import -c /etc/zfs/tank.cache tank` after unlocking the devices.

**Comment by brydling** [comment]
Thank you for your reply!

Is it really correct behaviour to fail mounting the root filesystem because another pool cannot be mounted yet?

I have tried various methods for decrypting the `tank` disks during the initrd phase (`clevis` and `decrypt_derived`), before importing the zfs pools to solve this problem, but can't get either of those methods to work. Dracut seems to figure out what disks make up the root filesystem, and only include these lines in the `/etc/crypttab` that gets included in the initrd image. In my example the `luks1 ...`-line is the only one that gets included. Even if I put the `initramfs` option on the `tank` disks they don't get included. Also, both these methods have their drawbacks even if I would get them to work, which made me fall for the solution of having my `tank` keys stored as files on `rpool`, This means initrd would have to do:

1. Decrypt `rpool` disks using `clevis`
2. Import `rpool`
3. Mount /
4. Decrypt `tank` disks using keys in /keys
5. Import `tank`

But then, if I could get initrd to execute steps 1, 2, 3 there would be no need to execute steps 4 and 5 because `tank` is totally unnecessary at this early stage in the boot process.

I can exclude the `tank` pool from the zpool.cache and then write a little separate service to import it automatically in the late boot, but I would still call that a workaround. I think it is a better workaround than my current one though, so I think I will try to do that. Thank you for the tip!

**Comment by iamjamestl** [comment]
@brydling, you are right that dracut tries to be smart and generate its own `/etc/crypttab` based on the disks that make up your root pool.  You can tell it to unlock and manage non-rpool encrypted disks by creating a file like `/etc/dracut.conf.d/10-crypt.conf` containing:

```
add_device+=" /dev/disk/by-uuid/f4dc141c-b9a4-4ca7-9b1f-333b80475113 /dev/disk/by-uuid/b8bcf96c-c29e-4241-af43-4b6759fa532d"
kernel_cmdline+=" rd.luks.uuid=f4dc141c-b9a4-4ca7-9b1f-333b80475113 rd.luks.uuid=b8bcf96c-c29e-4241-af43-4b6759fa532d"
``` 

where the UUIDs are from the output of `cryptsetup luksUUID /dev/sdwhatever`.  I use this in conjunction with a `/etc/crypttab` like:

```
nest-crypt0 UUID=f4dc141c-b9a4-4ca7-9b1f-333b80475113 /dev/disk/by-partlabel/key luks
nest-crypt1 UUID=b8bcf96c-c29e-4241-af43-4b6759fa532d /dev/disk/by-partlabel/key luks
```

to tell dracut to unlock non-root-pool devices at boot using a keyfile (in this case a key-partition labeled 'key').

It's probably not strictly proper to unlock non-root devices in the initramfs, but that depends on how critical you view the unlocked disks to the boot process.  In my case, I also use passphrases and letting dracut prompt me for a passphrase once and using that passphrase on all disks, rpool and not, is very nice.

**Comment by brydling** [comment]
@bunder2015 I tried enabling that service once, but it had a condition that the file `/etc/zfs/zpool.cache` doesn't exist. It refused to execute when the cache file was there. Of course that condition can be fixed too, but I already feel that I am too far from "normal" in my configuration. It feels like something is going to break with every `apt upgrade` from now on :)

@iamjamestl Thank you for the solution to that problem! I will try to get the `tank` disks to automatically unlock with `clevis` again using your dracut config. Though I would prefer having dracut figure all this out by itself. Since it by default already selectively decrypt just the disks needed to mount the root filesystem, it could also selectively import the pools needed and not import everything in the cache :)

The thing I don't think I get by exactly copying your solution is robustness. Instead of, as now, having the keys on a ZFS pool which can have redundancy, they will be on a single disk. Or is there any simple way to tell the system to try to reach the key in several different locations? It may be just a theoretical problem but when I remotely start the server and it drops to the emergency console because the disk with the keys has failed, I will be sad. But using `clevis` and a `tang` server I can decrypt them in the same way that I decrypt the root disks. The only drawback to that is that it is a bit slow, taking a few seconds for every `clevis` + `tang` unlock. I can live with 10 seconds extra boot time if it makes me sleep better :)

I will try it and let you know how it goes!

**Comment by brydling** [comment]
@iamjamestl 
I tried this just now. The disks now decrypt during initrd, but the `zfs-import-cache.service` fails with `The ZFS modules are not loaded.`.

If I load them manually using `/sbin/modprobe zfs` and then issue the command `/sbin/zpool import -c /etc/zfs/zpool.cache -aN` the `tank` pool is imported, but not `rpool`.

**Comment by brydling** [comment]
I am trying to work around this problem by removing `tank` from the default `/etc/zfs/zpool.cache`. It doesn't work for me. There is no way to set the property `cachefile` using `zpool set cachefile=none tank` and making it last after a reboot. If I run `zpool export tank` and then `zpool import tank`, the property is reset to `-` with SOURCE `default.`

Am I misunderstanding something or is this another bug in ZoL?

Trying to move from my old FreeBSD server to a new Linux server, but it doesn't go well for me.

**Comment by bunder2015** [comment]
This might be a little convoluted, but lets try this...  

Boot into a livecd and import -R /mnt rpool, set the cachefile to /etc/zfs/zpool.cache, copy it to /mnt/etc/zfs/zpool.cache.  Chroot into /mnt, and rebuild your initrd.  On the next boot it should only contain rpool and not tank.

**Comment by brydling** [comment]
@bunder2015 Wouldn't this be the same as running:

```
zpool set cachefile=none tank
rm /etc/zfs/zpool.cache
dracut -f
```
On my running system? Except that rpool will not be in the cachefile that way. However on my system, that is true anyway. `rpool` is never in the cachefile because `rpool` has `cachefile=none` (AND that property lasts reboots). I don't know why `rpool` has this property on my system. I followed this guide in setting up the system: https://github.com/zfsonlinux/zfs/wiki/Ubuntu-18.04-Root-on-ZFS . But I can't see that the guide explicitly sets `cachefile=none`

**Comment by brydling** [comment]
To clarify, have a look at this:

```
root@beta:/etc/zfs# ls -ltr
total 22
-rw-r--r-- 1 root root 9515 May  3  2018 zfs-functions
drwxr-xr-x 2 root root   13 Jan 26 15:22 zed.d
-rw-r--r-- 1 root root 1888 Feb 17 12:58 zpool.cache
root@beta:/etc/zfs# zpool get cachefile tank
NAME  PROPERTY   VALUE      SOURCE
tank  cachefile  -          default
root@beta:/etc/zfs# zpool set cachefile=/etc/zfs/zpool.tank.cache tank
root@beta:/etc/zfs# zpool get cachefile tank
NAME  PROPERTY   VALUE                      SOURCE
tank  cachefile  /etc/zfs/zpool.tank.cache  local
root@beta:/etc/zfs# ls -ltr
total 26
-rw-r--r-- 1 root root 9515 May  3  2018 zfs-functions
drwxr-xr-x 2 root root   13 Jan 26 15:22 zed.d
-rw-r--r-- 1 root root 1888 Feb 17 12:58 zpool.cache
-rw-r--r-- 1 root root 1888 Feb 17 12:59 zpool.tank.cache
root@beta:/etc/zfs# zpool export tank
root@beta:/etc/zfs# zpool import -c /etc/zfs/zpool.tank.cache -a
root@beta:/etc/zfs# zpool get cachefile tank
NAME  PROPERTY   VALUE      SOURCE
tank  cachefile  -          default
root@beta:/etc/zfs# ls -ltr
total 26
-rw-r--r-- 1 root root 9515 May  3  2018 zfs-functions
drwxr-xr-x 2 root root   13 Jan 26 15:22 zed.d
-rw-r--r-- 1 root root 1888 Feb 17 12:59 zpool.tank.cache
-rw-r--r-- 1 root root 1888 Feb 17 13:00 zpool.cache
```

As you see, the `cachefile`-property does not last export/import (or reboot). `tank` is added to the default `/etc/zfs/zpool.cache` on the execution of `zpool import -c /etc/zfs/zpool.tank.cache -a`.

**Comment by bunder2015** [comment]
I believe that is normal, even if you have a cachefile it will show up as `-` on the next import/boot.  I tried pinging @rlaager on irc to see if he has any insights that might help out.

**Comment by brydling** [comment]
@iamjamestl suggested this workaround in his first answer. Maybe he has some input too?

**Comment by rlaager** [comment]
> This whole ZFS-root + LUKS in Ubuntu feels extremely buggy at best.

It's definitely not the most stable thing. Hopefully we will have native encryption soon!

I'm not fully understanding the situation here. In the initramfs-tools case, the initrd imports rpool, so this can't happen. Is dracut not doing the same, but leaving it for systemd? I don't see how that works. That is, how do you even get the system booted far enough for that to be an option without a root filesystem. It must all happen in the initrd environment?

Anyway, I'm probably not going to be of much use here.

**Comment by brydling** [comment]
@rlaager it is in the initrd that rpool fails to import, because zfs-import-cache.service fails because my storage disks have not yet been decrypted in initrd. I decrypt these later. So the root problem is that rpool cannot be imported because there are others pools in the zpool cachefile. The other pools should not be required in initrd, but now they are. 

I have tried two workarounds for this problem :
1. Decrypt storage-disks in initrd (fails because of some bug in systemd-cryptsetup)
2. Remove tank-pool from zpool.cache (seems impossible if you read my thirdmost recent post)

I believe it was the problem in workaround #2 that @bunder2015 thought you could help with.

**Comment by rlaager** [comment]
Why are you using zpool-import-cache.service to import the root pool? Import the root pool specifically with “zpool import tool”, like initramfs does, and this issue goes away, right?

**Comment by brydling** [comment]
@rlaager I don't know. It seems like a bug and I thought ZoL was responsible for the behaviour. Or is it dracut entirely that is responsible for how the root pool is imported? @iamjamestl didn't seem to agree that it is a bug in his first answer.

**Comment by rlaager** [comment]
In initramfs land, the initrd is responsible for mounting the root filesystem. This is true whether it is ZFS or not. I'm not sure how dracut works, but surely it has to be similar. The whole point of an initrd is to run the system prior to the rootfs being mounted.

It sounds like you are saying you have systemd running in the initrd environment, with copies of things like zpool-import-cache.service. Why not just change this to specifically import the rpool? This could be any number of ways. You could have a separate service that just runs `zpool import rpool` and mounts the root filesystem (steps 2 & 3 above). Make that a dependency of step 4 and make step 4 a dependency of zpool-import-cache.service / zpool-import-scan.service, which would fill the role of step 5. Then everything runs in the right order, only once the step before it has worked, and everything is fine.

**Comment by brydling** [comment]
I am not familiar with either initramfs-tools or dracut, but yes it seems that dracut uses systemd. I am not sure how I can change the services that dracut uses, but I will investigate it. It seems like a bug that dracut tries to import all cached pools instead of just rpool, and I don't know where the bug report belongs. Is ZFS-integration in dracut the responsibility of ZoL or the responsibility of dracut?

**Comment by iamjamestl** [comment]
@byrdling, It's not a bug that dracut tries to import all of the pools in your zpool cache.  One of the major points of the cache is to describe what pools should be present on a system.  It has worked that way since the very beginning on Solaris.  The reason the `zpool cachefile` setting on tank changes after you've exported it is because "exporting" is meant for the system administrator to say "we're done with this pool."  You shouldn't export it.  Just `zpool set cachefile=/etc/zfs/tank.cache tank` and leave it alone.

With respect, these tools, zfs and its dracut module, are very mature, and they just work for typical configurations.  For example, I've been [abusing it since 2013](https://thestaticvoid.com/post/2013/10/26/how-i-do-encrypted-mirrored-zfs-root-on-linux/) and can now describe my advanced configuration in [under 100 lines of Puppet](https://github.com/iamjamestl/puppet-nest/blob/master/manifests/profile/base/dracut.pp).  What you are trying to achieve is a very advanced configuration and it will take absolute mastery of the [zfs dracut module](https://github.com/zfsonlinux/zfs/tree/master/contrib/dracut/90zfs), understanding of the difference between `zpool-import-cache` and `zpool-import-scan`, and when, why, and how they're invoked.  I don't think it's the fault of the tools that what you are trying to achieve is hard.  Again, with respect.  I think what you are trying to do is very cool and I really hope it works!

**Comment by iamjamestl** [comment]
@brydling, Here's what I would suggest for you.  Since in your original post you say you had the rpool-only configuration working with dracut and clevis, go back to that configuration.  Then, `zpool set cachefile=/etc/zfs/tank.cache tank`.  Then create a new systemd service to import tank like:

`/etc/systemd/systemd/zfs-import-tank.service`:
```
[Unit]
Description=Import tank by cache file
DefaultDependencies=no
Requires=systemd-udev-settle.service
After=systemd-udev-settle.service
After=cryptsetup.target
After=systemd-remount-fs.service
Before=dracut-mount.service
Before=zfs-import.target
ConditionPathExists=/etc/zfs/tank.cache

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStartPre=-/sbin/modprobe zfs
ExecStart=/sbin/zpool import -c /etc/zfs/tank.cache -aN

[Install]
WantedBy=zfs-import.target
```

And enable it with `systemctl enable zfs-import-tank`.

The end result should be that dracut unlocks, imports, and mounts your rpool, and systemd post-dracut runs `zfs-import-tank` before `zfs-import.target` to ensure your other zpool is imported.  Because `zfs-import-tank` depends on `cryptsetup.target`, its vdevs should be available and can use a key from your root pool because that is made available by dracut.

**Comment by brydling** [comment]
@iamjamestl Thank you for your input! Absolutely no offense taken.

We're discussing two separate problems I'm experiencing here, so I'll divide my answer:

**1.**
That dracut tries to import all pools in the default cachefile may not be a bug then, but I think that failing to mount rootfs because all pools cannot be imported is definitely a bug. Let's play with this scenario: I insert a USB-stick with ZFS and import that pool. While having my USB-stick inserted, I run "apt upgrade" which upgrades some packages that force an initrd-update. Now my new initrd-image includes a ZFS cachefile that contains my USB-pool. My machine is now un-bootable without the USB-stick inserted.

**2.**
Ok, so zpool export makes the cachefile-property reset. That makes sense. However, the same thing happens when I reboot my machine. I just tested:

```
zpool set cachefile=/etc/zfs/tank.cache tank
reboot
```

After the reboot, I now have this again:

```
root@beta:~# zpool get cachefile
NAME   PROPERTY   VALUE      SOURCE
rpool  cachefile  none       local
tank   cachefile  -          default
```

**Comment by iamjamestl** [comment]
1. I think you are right that in that particular scenario the USB pool would be shoved into your initramfs cache, but it is not my experience that missing pools cause dracut to fail to mount /.  I just tried this out with a USB pool and, though it spit out a failure message, dracut still succeeded to mount /:

```
Feb 18 14:14:37 eagle zfs-generator: starting
Feb 18 14:14:37 eagle zfs-generator: loading Dracut library from /lib/dracut-lib.sh
Feb 18 14:14:37 eagle zpool[3974]: cannot import 'test': no such pool or dataset
Feb 18 14:14:37 eagle zpool[3974]:         Destroy and re-create the pool from
Feb 18 14:14:37 eagle zpool[3974]:         a backup source.
Feb 18 14:14:37 eagle systemd[1]: zfs-import-cache.service: Main process exited, code=exited, status=1/FAILURE
Feb 18 14:14:37 eagle systemd[1]: zfs-import-cache.service: Failed with result 'exit-code'.
Feb 18 14:14:37 eagle systemd[1]: Failed to start Import ZFS pools by cache file.
Feb 18 14:14:37 eagle kernel: zpool (3974) used greatest stack depth: 12336 bytes left
Feb 18 14:14:37 eagle systemd[1]: Started Reload Configuration from the Real Root.
Feb 18 14:14:37 eagle systemd[1]: Reached target Initrd File Systems.
Feb 18 14:14:37 eagle systemd[1]: Reached target Initrd Default Target.
Feb 18 14:14:37 eagle systemd[1]: Starting dracut mount hook...
Feb 18 14:14:37 eagle dracut-mount[4319]: ZFS: No sysroot.mount exists or zfs-generator did not extend it.
Feb 18 14:14:37 eagle dracut-mount[4319]: ZFS: Mounting root with the traditional mount-zfs.sh instead.
Feb 18 14:14:37 eagle dracut-mount[4319]: ZFS: Using eagle/ROOT/gentoo as root.
Feb 18 14:14:37 eagle dracut-mount[4319]: ZFS: Mounting dataset eagle/ROOT/gentoo...
Feb 18 14:14:37 eagle kernel:  zd16: p1 p2 p3 p4
Feb 18 14:14:37 eagle systemd[1]: Started dracut mount hook.
```

2. Looks like you are right about that.  I had not tested that.  You need to explicitly preserve the cachefile location when you import the pool:

```
zpool import -c /etc/zfs/tank.cache -o cachefile=/etc/zfs/tank.cache tank
```

The `zfs-import-tank` service I suggested could be modified to add `-o cachefile=/etc/zfs/tank.cache` to the ExecStart command.  Maybe that behavior is unexpected.  Could be worth a discussion in a separate issue (this one is already a little long/confusing).


Regarding your other point, tank should have been removed from `/etc/zfs/zpool.cache` when you set `cachefile` to something else.  Use `zdb -U /path/to/cachefile` to list the contents of the cache to pin that behavior down.

**Comment by iamjamestl** [comment]
I'll just make another point.  Due to your complicated configuration, if you trust all of the disks that are attached or could be attached to your system, you could use `zfs-import-scan` instead of `zfs-import-cache` and eliminate all of the cachefiles completely.  Just set `cachefile=none` on all of your pools, ensure that `/etc/zfs/zpool.cache` is gone, rebuild dracut, and systemd will take care of the rest.  The `zfs-import-scan` service will import all pools visible under /dev, basically.

**Comment by brydling** [comment]
@iamjamestl 

1.

Your machine behaves better than mine then :) The problem we should be discussing then is why my system fails to mount rootfs because zfs-import-cache.service fails. See the boot log and the status from zfs-import-cache.service and rootfs.mount in my original problem report.

2.

Ah, ok! Yes that was a bit unexpected. Shall I create an issue for this?

More details:

```
root@beta:/etc/zfs# zpool get cachefile
NAME   PROPERTY   VALUE      SOURCE
rpool  cachefile  none       local
tank   cachefile  -          default
root@beta:/etc/zfs# ls -l
total 22
drwxr-xr-x 2 root root   13 Jan 26 15:22 zed.d
-rw-r--r-- 1 root root 9515 May  3  2018 zfs-functions
-rw-r--r-- 1 root root 1888 Feb 18 20:08 zpool.cache
root@beta:/etc/zfs# zdb -U /etc/zfs/zpool.cache
tank:
    version: 5000
    name: 'tank'
    state: 0
    txg: 93483
    pool_guid: 15972672765407542026
    errata: 0
    hostname: 'beta'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 15972672765407542026
        children[0]:
            type: 'mirror'
            id: 0
            guid: 2360523549362120705
            metaslab_array: 132
            metaslab_shift: 34
            ashift: 12
            asize: 3000027840512
            is_log: 0
            create_txg: 4
            com.delphix:vdev_zap_top: 129
            children[0]:
                type: 'disk'
                id: 0
                guid: 11211037769758641542
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 130
            children[1]:
                type: 'disk'
                id: 1
                guid: 15086454593332929620
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 131
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
root@beta:/etc/zfs# zpool set cachefile=/etc/zfs/tank tank
root@beta:/etc/zfs# zpool get cachefile
NAME   PROPERTY   VALUE          SOURCE
rpool  cachefile  none           local
tank   cachefile  /etc/zfs/tank  local
root@beta:/etc/zfs# ls -l
total 26
-rw-r--r-- 1 root root 1888 Feb 18 20:33 tank
drwxr-xr-x 2 root root   13 Jan 26 15:22 zed.d
-rw-r--r-- 1 root root 9515 May  3  2018 zfs-functions
-rw-r--r-- 1 root root 1888 Feb 18 20:08 zpool.cache
root@beta:/etc/zfs# zdb -U /etc/zfs/zpool.cache
tank:
    version: 5000
    name: 'tank'
    state: 0
    txg: 93483
    pool_guid: 15972672765407542026
    errata: 0
    hostname: 'beta'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 15972672765407542026
        children[0]:
            type: 'mirror'
            id: 0
            guid: 2360523549362120705
            metaslab_array: 132
            metaslab_shift: 34
            ashift: 12
            asize: 3000027840512
            is_log: 0
            create_txg: 4
            com.delphix:vdev_zap_top: 129
            children[0]:
                type: 'disk'
                id: 0
                guid: 11211037769758641542
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 130
            children[1]:
                type: 'disk'
                id: 1
                guid: 15086454593332929620
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 131
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
root@beta:/etc/zfs# zdb -U /etc/zfs/tank
tank:
    version: 5000
    name: 'tank'
    state: 0
    txg: 93483
    pool_guid: 15972672765407542026
    errata: 0
    hostname: 'beta'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 15972672765407542026
        children[0]:
            type: 'mirror'
            id: 0
            guid: 2360523549362120705
            metaslab_array: 132
            metaslab_shift: 34
            ashift: 12
            asize: 3000027840512
            is_log: 0
            create_txg: 4
            com.delphix:vdev_zap_top: 129
            children[0]:
                type: 'disk'
                id: 0
                guid: 11211037769758641542
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N0RUUCFY-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 130
            children[1]:
                type: 'disk'
                id: 1
                guid: 15086454593332929620
                path: '/dev/mapper/luks-ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5DC8C4C-part1'
                whole_disk: 0
                create_txg: 4
                com.delphix:vdev_zap_leaf: 131
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
```

The suggestion about zfs-import-scan.service. Does that import with cachefile=none in the import command, or will it make all pools re-appear in /etc/zfs/zpool.cache?

**Comment by brydling** [comment]
I just tested using `zfs-import-scan.service` instead. It works fine and is the easiest workaround so far in my opinion. However that service has a condition that /etc/zfs/zpool.cache does NOT exist, so I will have to be careful not to create an /etc/zfs/zpool.cache in the future or else my system will become un-bootable again. It is very easy to create that file by mistake because a regular `zpool import` without `-o cachefile=...` will create it.

**Comment by iamjamestl** [comment]
So, acknowledging that `zfs-import-scan` is not really the most ideal way to import pools, we're back to why your system fails to boot when `zfs-import-cache` fails.  I've just tried on my system with a bunch of pool configurations and boot parameters (root=, root=zfs:, root=zfs:&lt;dataset&gt;, etc.) and they all proceed to work even if some pools fail to import.  I've gone back and read your log messages in more detail and based on what you posted, my best guess is that your `rpool` wasn't included in the cache file in the initramfs at the time.  In fact, I think it's still not there because you posted:

```
> zdb -U /etc/zfs/zpool.cache
tank:
    version: 5000
    name: 'tank'
    state: 0
    txg: 93483
    pool_guid: 15972672765407542026
    errata: 0
    hostname: 'beta'
...
```
(no rpool)

Run `zpool set cachefile='' rpool` and rebuild your initramfs.  That should also solve your problem with the apparent stale cache file.

If it still doesn't work, then can you please post your kernel cmdline, any relevant zpool `bootfs` properties, `/etc/fstab` (if it contains entries related to your rpool)?  Rebuild your initramfs with `dracut -f -I /sbin/zdb` and if it fails to boot, provide the output of `zpool list` and `zdb`.

**Comment by brydling** [comment]
@iamjamestl I tried making a zpool.cache including rpool and generating a new initramfs-image, and it worked! Thank you!

Actually every working initramfs I have generated before this has been WITHOUT zpool.cache. It seems dracut imports rpool using some other means if zpool.cache does not exist. That other way is maybe a command using the -R option, because according to the man-page that will set cachefile=none. I had cachefile=none for rpool as you see in my previous listings.

So it seems dracut can successfully import rpool when no zpool.cache is present in initramfs. But then as soon as you create a zpool.cache, by adding other pools to the system, and regenerating initramfs, your system will not boot anymore. So this seems to not have been related to my tank-drives being encrypted at all. Even if zfs-import-cache would have successfully imported tank, it wouldn't have booted anyway because rpool was not in the zpool.cache.

So, how did my system end up in this state? I guess it was because of this guide: https://github.com/zfsonlinux/zfs/wiki/Ubuntu-18.04-Root-on-ZFS . It uses the -R flag in the zpool create command, which will set cachefile=none. This probably works fine with initramfs-tools, as @rlaager said previously because initramfs-tools imports rpool using a specific command and not using the cachefile even if one exists, but if you later switch to dracut you will get the problem I had.

Maybe it is better if dracut always imports rpool using a specific command for it, as initramfs-tools does? And then runs zfs-import-cache after that for trying to import all other pools? I see little point in having the fallback-method of importing rpool without a zpool.cache that dracut seems to have now, if it will lead to an un-bootable system later when you add other pools anyway.
